### ISSUE TITLE
ACT model compute flops

### ISSUE THREAD

**Issue by OBVIOUSDAWN** [gaopengcuhk/SMCA-DETR#23]
Dir sir,
thanks for your work on ACT model in transformer.I tried to integrate the ACT module into the task and modify the network following the pattern of [SMCA-DETR](https://github.com/gaopengcuhk/SMCA-DETR)/Adaptive_Cluster_Transformer/.My task is an object detection task, modified based on DETR.After act module was added, the overall training time of the task decreased by 2 hours, and the decrease of mAP just declare 0.6 as the paper say.However, when compute FLOPS, the total FLOPS on the network increased after ACT module added. I think it may be my compute code error. My flop calculation code is modified based on the original detr FLOPS calculation code, I don't know if there is any extra consideration for attention calculation.I would appreciate it if you could share information about the FLOP calculation code for the ACT module or point out what went wrong with my calculation code.

**Comment by OBVIOUSDAWN** [comment]
my compute flops code in [https://github.com/OBVIOUSDAWN/conpute_flops](url).Thank you

**Comment by minghangz** [comment]
Thank you for your interest in our work. We use this tools to computs FLOPs.
https://github.com/sovrasov/flops-counter.pytorch

**Comment by OBVIOUSDAWN** [comment]
> 感谢您对我们的工作感兴趣。我们使用这个工具来计算 FLOPs。 https://github.com/sovrasov/flops-counter.pytorch

Thanks for your help.I tried this program and got the GMac about whole model,but i found this program couldnt understand attention except "MultiheadAttention" ,so the "AdaMultiheadAttention" become 0 in all numbers.should i add anything in flops-counter to make it understand AdaMultiheadAttention or it shouldnt be calculate in this model?I would be appreciate it for your reaponse.

the results are as follows:
`(self_attn): AdaMultiheadAttention(
            0, 0.000% Params, 0.0 Mac, 0.000% MACs, 
            (out_proj): _LinearWithBias(0, 0.000% Params, 0.0 Mac, 0.000% MACs, in_features=256, out_features=256, bias=True)
            (attention): AdaClusteringAttention(
              0, 0.000% Params, 0.0 Mac, 0.000% MACs, 
              (dropout): Dropout(0, 0.000% Params, 0.0 Mac, 0.000% MACs, p=0.1, inplace=False)
              (softmax): WeightedSoftMax(0, 0.000% Params, 0.0 Mac, 0.000% MACs, )`

and
`(self_attn): MultiheadAttention(
            263.17 k, 0.633% Params, 31.54 MMac, 0.520% MACs, 
            (out_proj): _LinearWithBias(0, 0.000% Params, 0.0 Mac, 0.000% MACs, in_features=256, out_features=256, bias=True)
          )
          (multihead_attn): MultiheadAttention(
            263.17 k, 0.633% Params, 34.48 MMac, 0.569% MACs, 
            (out_proj): _LinearWithBias(0, 0.000% Params, 0.0 Mac, 0.000% MACs, in_features=256, out_features=256, bias=True)
          )`

**Comment by minghangz** [comment]
You should add a custom module hook which computs the MACs of AdaMultiheadAttention module ([here](https://github.com/sovrasov/flops-counter.pytorch/blob/7fe7d260bbdde12525c9bf109599683f5313a9a1/ptflops/pytorch_ops.py) are some references). 
Then pass in your custom hook like this:

```python
custom_modules_hooks = {
    MultiheadAttention: YOUR_HOOK,
}
get_model_complexity_info(model, input_res, custom_modules_hooks=custom_modules_hooks)
```

**Comment by OBVIOUSDAWN** [comment]
> You should add a custom module hook which computs the MACs of AdaMultiheadAttention module ([here](https://github.com/sovrasov/flops-counter.pytorch/blob/7fe7d260bbdde12525c9bf109599683f5313a9a1/ptflops/pytorch_ops.py) are some references). Then pass in your custom hook like this:
> 
> ```python
> custom_modules_hooks = {
>     MultiheadAttention: YOUR_HOOK,
> }
> get_model_complexity_info(model, input_res, custom_modules_hooks=custom_modules_hooks)
> ```

Thank you for your help. Before receiving your reply I tried to write a program to calculate the FPS about model detect on the same device. At the baseline model, the increase of FPS is significant after the ACT module added, but when combined with other modules, the increase of FPS is not significant and mAP decreases slightly. I think it may because other modules have greatly reduced the number of inputs before encoder that it cannot perform better after adding ACT. At the same time, I noticed that MTKD was used for distillation of detr in the paper of ACT module. I have never seen the distillation of transformer in detr before, and I want to know what kind of distillation means by MKTD. I have seen the distillation network introduced in Figure5. Is a normal transformer parallel with a transformer with act added?

Besides, I would really appreciate it if we can communicate in Chinese.As a beginner, I don't have enough understanding of many tasks, so there may be some errors if I describe them in English. Thank you very much for your help and I look forward to your reply.
